roachtest: tpccbench/nodes=12/cpu=16 failed

[(roachtest).tpccbench/nodes=12/cpu=16 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1683224&tab=buildLog) on [master@413fd1e46c2936029458db495efa242acb1b7d52](https://github.com/cockroachdb/cockroach/commits/413fd1e46c2936029458db495efa242acb1b7d52):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200110-1683224/tpccbench/nodes=12/cpu=16/run_1
	test_runner.go:731: test timed out (10h0m0s)
```

<details><summary>Repro</summary><p>


Artifacts: [/tpccbench/nodes=12/cpu=16](https://teamcity.cockroachdb.com/viewLog.html?buildId=1683224&tab=artifacts#/tpccbench/nodes=12/cpu=16)
```
make stressrace TESTS=tpccbench/nodes=12/cpu=16 PKG=./pkg/roachtest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpccbench/nodes=12/cpu=16 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1696119&tab=buildLog) on [master@96d7e50acfaa63b5b60addd7cec8e750a4a4d333](https://github.com/cockroachdb/cockroach/commits/96d7e50acfaa63b5b60addd7cec8e750a4a4d333):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200117-1696119/tpccbench/nodes=12/cpu=16/run_1
	cluster.go:2201,tpcc.go:720,tpcc.go:561,test_runner.go:716: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2192
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2200
		    main.runTPCCBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:720
		    main.registerTPCCBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:561
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:716
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 3: dead
```

<details><summary>Repro</summary><p>


Artifacts: [/tpccbench/nodes=12/cpu=16](https://teamcity.cockroachdb.com/viewLog.html?buildId=1696119&tab=artifacts#/tpccbench/nodes=12/cpu=16)
```
make stressrace TESTS=tpccbench/nodes=12/cpu=16 PKG=./pkg/roachtest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/dist_sender_test.go
+++ b/pkg/kv/dist_sender_test.go
@@ -2737,7 +2737,7 @@ func TestMultipleErrorsMerged(t *testing.T) {
 		{
 			err1:   abortErr,
 			err2:   conditionFailedErr,
-			expErr: "unexpected value",
+			expErr: "TransactionAbortedError(ABORT_REASON_ABORTED_RECORD_FOUND)",
 		},
 		{
 			err1:   conditionFailedErr,

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -779,6 +779,11 @@ func (tc *TxnCoordSender) updateStateLocked(
 			PrevError: pErr.String(),
 		})
 		tc.mu.txn.Update(errTxn)
+		if errTxn.Status != roachpb.PENDING {
+			// We only expect TransactionAbortedError to carry an aborted txn.
+			log.Errorf(ctx, "programming error: ABORTED txn in error: %s. txn: %s", pErr, errTxn)
+			tc.cleanupTxnLocked(ctx)
+		}
 	}
 	return pErr
 }

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -63,14 +63,22 @@ const (
 	// ErrorScoreTxnRestart indicates that the transaction should be restarted
 	// with an incremented epoch.
 	ErrorScoreTxnRestart
-	// ErrorScoreTxnAbort indicates that the transaction is aborted. The
-	// operation can only try again under the purview of a new transaction.
-	ErrorScoreTxnAbort
 	// ErrorScoreNonRetriable indicates that the transaction performed an
 	// operation that does not warrant a retry. Often this indicates that the
 	// operation ran into a logic error. The error should be propagated to the
 	// client and the transaction should terminate immediately.
 	ErrorScoreNonRetriable
+	// ErrorScoreTxnAbort indicates that the transaction is aborted. The
+	// operation can only try again under the purview of a new transaction.
+	//
+	// This error has the highest priority because, as far as KV is concerned, a
+	// TransactionAbortedError is impossible to recover from (whereas
+	// non-retriable errors could conceivably be recovered if the client wanted to
+	// ignore them). Also, the TxnCoordSender likes to assume that a
+	// TransactionAbortedError is the only way it finds about an aborted
+	// transaction, and so it benefits from all other errors being merged into a
+	// TransactionAbortedError instead of the other way around.
+	ErrorScoreTxnAbort
 )
 
 // ErrPriority computes the priority of the given error.

--- a/pkg/sql/distsql_running_test.go
+++ b/pkg/sql/distsql_running_test.go
@@ -228,17 +228,6 @@ func TestDistSQLReceiverErrorRanking(t *testing.T) {
 			expErr: "TransactionRetryWithProtoRefreshError: TransactionRetryError",
 		},
 		{
-			// A TransactionAbortedError overwrites another retriable one.
-			err:    abortErr,
-			expErr: "TransactionRetryWithProtoRefreshError: TransactionAbortedError",
-		},
-		{
-			// A non-aborted retriable error does not overried the
-			// TransactionAbortedError.
-			err:    retryErr,
-			expErr: "TransactionRetryWithProtoRefreshError: TransactionAbortedError",
-		},
-		{
 			// A non-retriable error overwrites a retriable one.
 			err:    fmt.Errorf("err1"),
 			expErr: "err1",
@@ -247,6 +236,17 @@ func TestDistSQLReceiverErrorRanking(t *testing.T) {
 			// Another non-retriable error doesn't overwrite the previous one.
 			err:    fmt.Errorf("err2"),
 			expErr: "err1",
+		},
+		{
+			// A TransactionAbortedError overwrites anything.
+			err:    abortErr,
+			expErr: "TransactionRetryWithProtoRefreshError: TransactionAbortedError",
+		},
+		{
+			// A non-aborted retriable error does not overried the
+			// TransactionAbortedError.
+			err:    retryErr,
+			expErr: "TransactionRetryWithProtoRefreshError: TransactionAbortedError",
 		},
 	}
 

